docs/proposals/component-registration.md

```diff
@@ -0,0 +1,320 @@
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Proposal: Component Registration
+
+## Table of Contents
+
+<!-- BEGIN MUNGE: GENERATED_TOC -->
+
+- [Proposal: Component Registration](#proposal-component-registration)
+  - [Table of Contents](#table-of-contents)
+  - [Terminology](#terminology)
+  - [Abstract](#abstract)
+  - [Problems](#problems)
+    - [ComponentStatuses](#componentstatuses)
+    - [Endpoints](#endpoints)
+  - [Use Cases](#use-cases)
+    - [Deployment](#deployment)
+    - [Debugging](#debugging)
+    - [Notification](#notification)
+    - [Reporting](#reporting)
+    - [Self-Healing](#self-healing)
+  - [Design](#design)
+    - [API Changes](#api-changes)
+      - [API Types](#api-types)
+    - [Kubectl (CLI)](#kubectl-cli)
+    - [Storage](#storage)
+    - [Status Updates](#status-updates)
+  - [Example](#example)
+  - [Alternate Naming](#alternate-naming)
+  - [Next Steps](#next-steps)
+
+<!-- END MUNGE: GENERATED_TOC -->
+
+## Terminology
+
+A Kubernetes Cluster **Component**...
+
+- Is considered a part of the cluster
+- Must register (or be registered) with Kubernetes (apiserver) to join the cluster
+- Must know about Kubernetes (speak the API)
+- Must be deployed/managed by the cluster admin/operator (rather than a user)
+- Should expose a method for remotely probing the liveness and readiness of each instance (e.g. tcp/http/https).
+- May be scheduled/deployed on the cluster (primary components vs addon components)
+- May be required for the cluster to function (required vs optional)
+- May be added/removed from a running cluster (mutable cluster component membership over time)
+- May require admin API access (privileged behavior)
+- May consist of one or more instances
+- May be deployed on different (network accessible) machines
+- May be upgraded over time
+
+The **Primary Components** currently include apiserver, controller-manager, and scheduler. The lifecycle of these components is managed by some external system. These components are required for Kubernetes to function.
+
+The **Addon Components** currently include kube-dns and kube-ui. The lifecycle of these components is managed by the kubernetes cluster that they provide functionality to. These components add functionality, but are not required for Kubernetes to function. Kubernetes currently ships with some of these required for "conformance", but they're technically not required for minimal orchestration and scheduling.
+
+The"**Hyperkube** is a single binary that can be run as each of the three primary component binaries. Hyperkube may also refer to the set (or pod) of primary component containers or the (virtual) machine that those containers run on. Not all Kubernetes deployments use or should be required to use the Hyperkube.
+
+**Master** has been sometimes used in reference to both the APIServer component and the (virtual) machine which hosts the three primary components. I'm going to avoid this term, because it is ambiguous.
```

Comment:
+1000 to avoid it in any technical specs, docs, etc... until its qualified and fully defined.